R/as-data-frame.R

```diff
@@ -1,34 +1,64 @@
-CrunchDataFrame <- function (dataset) {
+CrunchDataFrame <- function (dataset, order) {
     ## S3 constructor method for CrunchDataFrame. terms.formula doesn't seem
     ## to like S4 subclass of environment
     stopifnot(is.dataset(dataset))
     out <- new.env()
     out$.crunchDataset <- dataset
+    out$.names <- c()
+    # set the order of the dataset based on order function, if no order is 
+    # given return all rows in the dataset in the order they appear
+    if (is.null(order)) {
+        out$.order <- order
+    } else {
+        out$.order <- order
+    }
+    
     with(out, {
         ## Note the difference from as.environment: wrapped in as.vector
-        for (a in aliases(allVariables(dataset))) {
-            eval(substitute(delayedAssign(v, as.vector(.crunchDataset[[v]])),
-                list(v=a)))
+        for (.a in aliases(allVariables(dataset))) {
+            eval(substitute(delayedAssign(v, {
+                if (is.null(.order)) {
+                    as.vector(.crunchDataset[[v]])
+                } else {
+                    # as.numeric(factor(.order)) gets the dense ranking of the 
+                    # ordering so that the non-duplicated results returned in 
+                    # crunch order can be used in the order specified in .order.
+                    as.vector(.crunchDataset[[v]][.order])[as.numeric(factor(.order))]
```

Comment:
This is probably a hack, but was the only way I could find to do what I needed. (Looking back, and digging through some source of [`dense_rank`](https://github.com/tidyverse/dplyr/blob/master/R/rank.R#L72-L73) we really probably just need `match(r, sort(unique(r)))` here instead which is much clearer, at least to me.) When we send subsets by rownumber to the server the server actually returns them ordered like they were in the dataset not like they were requested. But what is returned is only the rows that were selected, so we can't use `.order` here again to match the ordering of the rows that are returned, because they frequently will not be the correct indices for what is returned.

But ultimately what the `[as.numeric(factor(.order))]`|`[match(.order, sort(unique(.order)))]` is doing is correcting the `.order` indices to match what the API returns. So if we request rows `c(12, 12, 1, 2, 3, 4, 4, 5, 5, 5, 6, 6, 6, NA)` we will order the responses we get back as `c(7, 7, 1, 2, 3, 4, 4, 5, 5, 5, 6, 6, 6, NA)`

We might consider adding this on to the `[` method of variables so that the r-client always overrides the API behavior, and returns what an R user would expect in those contexts.